File: bench/thumbhash-py/bench.py
"""Benchmark the PyPI `thumbhash` package on the same 100x100 gradient
input. Output: NDJSON to stdout.
"""
from __future__ import annotations

import json
import statistics
import time

import thumbhash


def gradient_rgba(w: int, h: int) -> list[int]:
    rgba = bytearray(w * h * 4)
    for y in range(h):
        for x in range(w):
            p = (y * w + x) * 4
            rgba[p] = round(x * 255 / max(w - 1, 1))
            rgba[p + 1] = round(y * 255 / max(h - 1, 1))
            rgba[p + 2] = min(255, int((x + y) * 0.3))
            rgba[p + 3] = 255
    return list(rgba)


def measure(fn, warmup, iters):
    for _ in range(warmup):
        fn()
    samples = []
    for _ in range(iters):
        t0 = time.perf_counter_ns()
        fn()
        samples.append((time.perf_counter_ns() - t0) / 1000.0)
    samples.sort()
    return samples[len(samples) // 2], samples[int(len(samples) * 0.95)], samples[0]


def report(mode, op, w, h, s, iters, extra=None):
    out = {
        "impl": "py-thumbhash",
        "mode": mode,
        "op": op,
        "w": w,
        "h": h,
        "median_us": round(s[0], 2),
        "p95_us": round(s[1], 2),
        "min_us": round(s[2], 2),
        "iters": iters,
        "mpix_per_s": round((w * h) / s[0], 3),
    }
    if extra:
        out.update(extra)
    print(json.dumps(out), flush=True)


def main():
    w, h = 100, 100
    rgba = gradient_rgba(w, h)
    hash_ = {"v": []}
    s = measure(lambda: hash_.__setitem__("v", thumbhash.rgba_to_thumb_hash(w, h, rgba)), 30, 200)
    report("dct", "encode", w, h, s, 200, {"hash_bytes": len(hash_["v"])})

    hb = hash_["v"]
    s = measure(lambda: thumbhash.thumb_hash_to_rgba(hb), 10, 50)
    report("dct", "decode_default", w, h, s, 50)


if __name__ == "__main__":
    main()

File: bench/thumbhash-js/bench.mjs
// Benchmark thumbhash (official JS) — same workload as Rust/Python/Go.
// Output: NDJSON to stdout.

import { rgbaToThumbHash, thumbHashToRGBA } from "thumbhash";

function gradient(w, h) {
  const rgba = new Uint8Array(w * h * 4);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const p = (y * w + x) * 4;
      rgba[p] = Math.round((x * 255) / Math.max(w - 1, 1));
      rgba[p + 1] = Math.round((y * 255) / Math.max(h - 1, 1));
      rgba[p + 2] = Math.min(255, Math.floor((x + y) * 0.3));
      rgba[p + 3] = 255;
    }
  }
  return rgba;
}

function measure(fn, warmup, iters) {
  for (let i = 0; i < warmup; i++) fn();
  // Node hrtime() resolution on Windows is good (~100ns), but for very fast
  // ops we still batch for stability.
  const batch = 50;
  const samples = [];
  for (let i = 0; i < iters; i++) {
    const t0 = process.hrtime.bigint();
    for (let j = 0; j < batch; j++) fn();
    const dt_ns = Number(process.hrtime.bigint() - t0);
    samples.push(dt_ns / batch / 1000.0);
  }
  samples.sort((a, b) => a - b);
  return {
    median_us: samples[Math.floor(samples.length / 2)],
    p95_us: samples[Math.floor(samples.length * 0.95)],
    min_us: samples[0],
    iters,
  };
}

function report(mode, op, w, h, s, extra = {}) {
  const out = {
    impl: "js",
    mode,
    op,
    w,
    h,
    median_us: Math.round(s.median_us * 100) / 100,
    p95_us: Math.round(s.p95_us * 100) / 100,
    min_us: Math.round(s.min_us * 100) / 100,
    iters: s.iters,
    mpix_per_s: Math.round(((w * h) / s.median_us) * 1000) / 1000,
    ...extra,
  };
  console.log(JSON.stringify(out));
}

function main() {
  const w = 100, h = 100;
  const rgba = gradient(w, h);
  let hash;
  let s = measure(() => { hash = rgbaToThumbHash(w, h, rgba); }, 30, 200);
  report("dct", "encode", w, h, s, { hash_bytes: hash.length });

  s = measure(() => { thumbHashToRGBA(hash); }, 10, 50);
  report("dct", "decode_default", w, h, s);
}

main();
